accio-cache/src/main/java/io/accio/cache/CacheManager.java

```diff
@@ -250,29 +265,38 @@ public void removeTempFile(PathInfo pathInfo)
         }
     }
 
-    public TaskInfo createTaskUtilDone(AccioMDL mdl)
+    public List<TaskInfo> createTaskUtilDone(AccioMDL mdl)
+    {
+        return createTask(mdl)
+                .thenApply(taskInfos -> taskInfos.stream()
+                        .map(taskInfo -> {
+                            tasks.get(taskInfo.getCatalogSchemaTableName()).waitUntilDone();
+                            return getTaskInfo(taskInfo.getCatalogSchemaTableName()).join();
+                        })
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .collect(toList())).join();
+    }
+
+    public CompletableFuture<List<TaskInfo>> createTask(AccioMDL mdl)
     {
-        Optional<TaskInfo> taskInfoOptional = createTask(mdl)
-                .thenCompose(taskInfo -> {
-                    tasks.get(taskInfo.getTaskId()).waitUntilDone();
-                    return getTaskInfo(taskInfo.getTaskId());
-                })
-                .join();
-        return taskInfoOptional.orElseThrow(() -> new RuntimeException("Failed to create task"));
+        return supplyAsync(() ->
+                mdl.listCached().stream().map(cacheInfo -> createTask(mdl, cacheInfo).join()).collect(toList()));
     }
 
-    public CompletableFuture<TaskInfo> createTask(AccioMDL mdl)
+    public CompletableFuture<TaskInfo> createTask(AccioMDL mdl, CacheInfo cacheInfo)
     {
         return supplyAsync(() -> {
-            String taskId = randomUUID().toString();
-            TaskInfo taskInfo = new TaskInfo(taskId, mdl.getCatalog(), mdl.getSchema(), RUNNING, Instant.now());
-            Task task = new Task(taskInfo, refreshCache(mdl));
-            tasks.put(taskId, task);
+            CatalogSchemaTableName catalogSchemaTableName = new CatalogSchemaTableName(mdl.getCatalog(), mdl.getSchema(), cacheInfo.getName());
+            TaskInfo taskInfo = new TaskInfo(mdl.getCatalog(), mdl.getSchema(), cacheInfo.getName(), RUNNING, Instant.now());
+            // To fix flaky test, we pass value to tasks instead of a reference;
+            Task task = new Task(TaskInfo.copyFrom(taskInfo), handleCache(mdl, cacheInfo));
```

Comment:
A little confused: why`handleCache` calls `createTask`, and `createTask` also calls `handleCache`?

Reply:
> A little confused: why`handleCache` calls `createTask`, and `createTask` also calls `handleCache`?

`createTask` means create a new cache. To get the latest end time, we need to create a new task periodically. That's why `handleCache` will use it in L127.

Reply:
as discussed offline, since TestRefreshCache#testRefreshSingleModel need to assert if TaskInfo status is `RUNNING`, that's why we use copyFrom here. To avoid task status was updated in other places.